### PR TITLE
[FLINK-13513][ml] Add Mapper and related classes for later algorithms

### PR DESCRIPTION
**Pull request by xuyang1706**
## What is the purpose of the change

Add  the FlatMapper and related classes  for later algorithm implementations. Define the abstract classes for the single-thread executor for Row type data.

## Brief change log

  - *Add the definition of abstract class FlatMapper*
  - *Add interface MapOpInterface for  the map operation of Row type data.*
  - *Add abstract class FlatModelMapper, Mapper extends FlatMapper*
  - *Add abstract class ModelMapper extends FlatModelMapper*

## Verifying this change
This change added tests and can be verified as follows:

- run test case pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+	 * @throws Exception if {@link #flatMap(Row, Collector)} throws exception.
+	 */
+	public void flatMap(Iterable <Row> rows, Collector<Row> output) throws Exception {
+		for (Row row : rows) {
+			flatMap(row, output);
+		}
+	}
+
+	/**
+	 * Get the table schema(includs column names and types) of the calculation result.
```
Comment (weibo611) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java]: "includs" is a spell error

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+package org.apache.flink.ml.common.mapper;
+
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+
+/**
+ * Abstract class for mappers with model.
+ * ModelMapper transform one Row type data into one Row type result data.
```
Comment (weibo611) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: "transform" should be "transforms"

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java (hunk 1 above)
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java]: thx, it has been fixed

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java (hunk 2 above)
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: thx, it has been fixed

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+	 * @param modelRows the list of Row type data
+	 */
+	public abstract void loadModel(List <Row> modelRows);
+
+	/**
+	 * Generate new instance of given FlatModelMapper class without model data.
+	 * The instance can not deal with real data, but it could be used to get the output result schema.
+	 *
+	 * @param flatModelMapperClassName Name of the FlatModelMapper class
+	 * @param modelScheme              The schema of input Table type model.
```
Comment (walterddr) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java]: it is not immediately clear to me what's the difference between `modelScheme` &  `dataSchema`. 
Also why is modelScheme of the type `TableSchema` ?

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.types.Row;
+
+import java.util.List;
+
+/**
+ * Abstract class for flatMappers with model.
+ * FlatModelMapper transform one Row type data into zero, one, or more Row type result data.
```
Comment (walterddr) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java]: IMO, This is duplicate Javadoc similar to `FlatMapper`. Can we instead explain how the "model" part works. Basic `FlapMapper` already explains these.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/MapOpInterface.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+ */
+
+package org.apache.flink.ml.common.mapper;
+
+import org.apache.flink.types.Row;
+
+/**
+ * Interface for the map operation of Row type data.
+ */
+public interface MapOpInterface {
```
Comment (walterddr) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/MapOpInterface.java]: should this file be named just `MapOp`. There's already a `MatVecOp` class thought we would like to keep the naming conversion as well?

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+
+import java.io.Serializable;
+
+/**
+ * Abstract class for flatMappers.
+ * FlatMapper transform one Row type data into zero, one, or more Row type result data.
+ * Operations that produce multiple strictly one Row type result data per Row type data
```
[the remark this answers is not in the document]
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java]: Thanks, FlatMapper can not support multiple rows as its inputs and we have been removed the `Iterable<Row>` in this interface

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+	 */
+	public abstract void flatMap(Row row, Collector<Row> output) throws Exception;
+
+    /**
+	 * Wrapper method for the iterable input.
+	 * @param rows the input rows.
+	 * @param output the output collector
+	 * @throws Exception if {@link #flatMap(Row, Collector)} throws exception.
+	 */
+	public void flatMap(Iterable <Row> rows, Collector<Row> output) throws Exception {
```
[the remark this answers is not in the document]
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java]: Thanks, removed

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java (hunk 3 above)
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java]: Thanks, the model is the machine learning model that use the Table as its representation (serialized to Table from the memory or deserialized from Table to memory). Thus, the model need `modelSchema` and the predict data need `dataSchema`.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java (hunk 4 above)
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java]: Thanks, removed.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/MapOpInterface.java (hunk 5 above)
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/MapOpInterface.java]: Thanks for your advice, I renamed this interface to `MapOperable`.

**Comment by xuyang1706** [comment]
> Thanks for the contribution @xuyang1706 . I took a quick look and I think there are some confusions I have with the Javadocs. Can you please take a look and let me know if any of my questions make sense?

@walterddr , thanks for your comments and suggestion. I have refined the JavaDoc and add more descriptions, removed unnecessary method and renamed the interface name.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java (hunk 3 above)
Comment (walterddr) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java]: ah. I see now. I think the prhase "the schema of the input Table" part confused me.
I would probably rephrase it to
```
@param modelScheme    The model scheme represented in Table format
```

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+	 */
+	public abstract void loadModel(List<Row> modelRows);
+
+	/**
+	 * Generate new instance of given FlatModelMapper class without model data.
+	 * The instance can not deal with real data, but it could be used to get the output result schema.
+	 *
+	 * @param flatModelMapperClassName Name of the FlatModelMapper class
+	 * @param modelScheme              The schema of input Table type model.
+	 * @param dataSchema               The schema of input Table type data.
```
Comment (walterddr) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java]:
```
@param dataSchema    The schema of the input data represented in Table format
```

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+	/**
+	 * This method override the {@link FlatMapper#flatMap(Row, Collector)} to map a row
+	 * to a new row which collected to {@link Collector}.
+	 *
+	 * @param row the input row.
+	 * @param output the output collector
+	 * @throws Exception if {@link Collector#collect(Object)} throws exception.
+	 */
+	@Override
+	public void flatMap(Row row, Collector<Row> output) throws Exception {
```
Comment (ex00) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java]: it is same as in `ModelMapper`
I suppose this logic can be move to `FlatMapper` class directly, because in this version of code base all implementation this method is equal.

Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java]: Thanks for your advice. For One of the extended class `FlatModelMapper` not support the map() function, we can not override its flatMap() function. Thus, I can not move the function to  FlatMapper class.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java (hunk 8 above)
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java]: Thanks, changed.

**Comment by xuyang1706** [comment]
> thanks for the update @xuyang1706 and sorry for the late reply. Overall I think the change looks good to me. @becketqin do you have any additional comments?

Thanks @walterddr , your suggested descriptions are much better, I have rephrased them.

**Comment by xuyang1706** [comment]
> Hi @xuyang1706
> Thanks for you PR. I have left only one minor comment, please look it if you have time.
> In general this changes LGTM.
> Thanks!

Thanks @ex00 for your helpful comments and suggestion.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java (hunk 9 above)
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java]: Yes, there are other good ways to implement. I'd like to choose current one. Thanks for your helpful discussion and advice!

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+	/**
+	 * Generate new instance of given FlatMapper class.
+	 *
+	 * @param flatMapperClassName Name of the FlatMapper class
+	 * @param dataSchema          The schema of the input data represented in Table format.
+	 * @param params              The parameters for the instance construction.
+	 * @return new instance of given FlatMapper class
+	 * @throws Exception if flatMapperClass is not the class of {@link FlatMapper}
+	 */
+	public static FlatMapper of(
```
[the remark this answers is not in the document]
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java]: Thanks, changed the static method.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java [hunk 11]
```diff
@@ -0,0 +1,138 @@
...
+	 * The instance can not deal with real data, but it could be used to get the output result schema.
+	 *
+	 * @param flatModelMapperClassName Name of the FlatModelMapper class
+	 * @param modelScheme              The model scheme represented in Table format.
+	 * @param dataSchema               The schema of the input data represented in Table format.
+	 * @param params                   The parameters for the instance construction.
+	 * @return The object of {@link FlatModelMapper}
+	 * @throws Exception if flatModelMapperClass is not the class of {@link FlatModelMapper}
+	 */
+	public static FlatModelMapper of(
```
[the remark this answers is not in the document]
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java]: Thanks, changed the static method.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/MapOperable.java [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+package org.apache.flink.ml.common.mapper;
+
+import org.apache.flink.types.Row;
+
+/**
+ * Interface for the map operation of Row type data.
+ *
+ * <p>MapOperable maps a row to a new row.
+ */
+public interface MapOperable {
```
[the remark this answers is not in the document]
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/MapOperable.java]: Thanks, refactored the definition of Mapper.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+
+import org.apache.flink.ml.api.misc.param.Params;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.types.Row;
+import org.apache.flink.util.Collector;
+
+/**
+ * Abstract class for mappers.
+ */
+public abstract class Mapper extends FlatMapper implements MapOperable {
```
[the remark this answers is not in the document]
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java]: Thanks, refactored the definition of Mapper.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+		Class flatMapperClass,
+		TableSchema dataSchema,
+		Params params) throws Exception {
+
+		if (FlatMapper.class.isAssignableFrom(flatMapperClass)) {
+			return (FlatMapper) flatMapperClass.getConstructor(TableSchema.class, Params.class)
```
[the remark this answers is not in the document]
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java]: Thanks, changed.

**Comment by xuyang1706** [comment]
> @xuyang1706 Thanks for the patch. The structure and semantic in the patch seems a little confusing to me. I created a PR against your PR with my suggestions. The PR might be a little messy at this point. But should be clean once you rebase the `flatmapper` branch on master.
> 
> Please let me know what do you think. Thanks.

Thanks for your comments and discussion, @becketqin. We refactored the codes, please look it if you have time.  Thanks, -Xu

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+		/**
+		 * Load model data from flink's broadcast dataset.
+		 */
+		BroadcastDataset,
+
+		/**
+		 * Load model data from data rows.
+		 */
+		Rows
+	}
```
Comment (becketqin) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: It is probably worth separating out the model source so it can be reused by other classes. Also it seems indication that we may need two ModelSource implementations. I have created a PR against your PR branch with my suggestions.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+		super(dataSchema, params);
+		this.modelFieldNames = modelSchema.getFieldNames();
+		this.modelFieldTypes = modelSchema.getFieldDataTypes();
+	}
+
+	protected TableSchema getModelSchema() {
+		return TableSchema.builder().fields(this.modelFieldNames, this.modelFieldTypes).build();
+	}
+
+	public final void setModelBroadcastName(String modelBroadcastName) {
```
Comment (becketqin) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: Do we know the ModelSource at construction time? If so, the model source can be passed to the constructor to make the class immutable.

**Comment by becketqin** [comment]
Also, just curious what about the `FlatMapper` classes? Do we still want them?

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java (hunk 15 above)
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: Yes, it is better to separate out the model source.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java (hunk 16 above)
Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: Thanks, the model source is passed to the constructor.

**Comment by xuyang1706** [comment]
> Also, just curious what about the `FlatMapper` classes? Do we still want them?

Thanks for your advice, @becketqin. The `FlatMapper` class is removed from this PR, and will be submitted later with its related functionalities.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+	 * Load model from the list of Row type data.
+	 *
+	 * @param modelRows the list of Row type data
+	 */
+	public abstract void loadModel(List<Row> modelRows);
+
+	@Override
+	public final void open(Configuration parameters) {
+		Preconditions.checkState(this.modelSource != null, "model source not set.");
+		List<Row> rows = this.modelSource.getModelRows(getRuntimeContext());
```
Comment (walterddr) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: Shouldn't the parameters be part of the interface signature?

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+/**
+ * Abstract class for mappers.
+ */
+public abstract class Mapper extends RichMapFunction<Row, Row> implements Serializable {
+
+	/**
+	 * schema of the input.
+	 */
+	private final String[] dataFieldNames;
```
Comment (bowenli86) [inline flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java]: Why not just keep the dataSchema passed in from constructor? It seems a bit unintuitive to me

Comment (qiuxiafei) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java]: It not serializable while the mapper class will be serialized to TM for execution.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java (hunk 17 above)
Comment (qiuxiafei) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: Actually it shouldn't, we're working on an update which decouples ModelMapper and Mapper  from flink's `MapFunction`.

Comment (xuyang1706) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: We have updated the code.

**Comment by xuyang1706** [comment]
> Hi @xuyang1706 @becketqin Thanks for the discussion and sorry for the late response. This PR currently seems to be far away from the original intent of the [JIRA](https://issues.apache.org/jira/browse/FLINK-13513). I would suggest please change the title as well as the content for the commit message to make it more clear that we only support `ModelMapper` and not FlatMapper.

Thanks for your suggestion, @walterddr . I have changed the title.

File: flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java (hunk 17 above)
Comment (walterddr) [reply flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java]: I see. however at this current stage it seemed weird that the open method does not use `Configuration` passed in, instead using the more generic `getRuntimeContext`. 
I am find with current impl actually. thanks for the explanation.

**Comment by becketqin** [comment]
@xuyang1706 I created a PR with some minor suggestion. Most of them are just java docs. Can you check and see if you have any concerns. I'll merge the code if you are OK with the changes. Thanks.

Thanks for the review @walterddr @bowenli86 @ex00 @weibo611

**Comment by xuyang1706** [comment]
> @xuyang1706 I created a PR with some minor suggestion. Most of them are just java docs. Can you check and see if you have any concerns. I'll merge the code if you are OK with the changes. Thanks.
> 
> Thanks for the review @walterddr @bowenli86 @ex00 @weibo611

Thanks @becketqin , your changes are good to me, please merge this PR.

**Comment by walterddr** [comment]
close via [005bda9be](https://github.com/apache/flink/commit/005bda9be)
